Sprint20 Add Random Variables (Using Abstract Syntax Tree)

nabriis opened the pull request:
closes #202, #85, #29

Related to #250

# Tracking of final changes to PR
17/11/2023
- [x] Consider making name of random variable different than parameter name.
- [x] Update logic in to_cuqi_format (removed again)
- [x] Explain condition method and gradient method
- [x] Explain geometry and dim etc.

18/11/2023
- [ ] Consider input name to JointDistribution and BayesianProblem (components)
- [ ] Consider if force_ndarray should detect distribution and give useful error message (potential common mistake), as well as Model.forward(Distribution)
- [ ] Consdier if it should be par_name for Distribution. Maybe another name?
- [ ] Disable `condition` related properties of RVs for now.
- [ ] Remove `__call__` method of Distribution. Rename to e.g. `fix` or `given`
- [x] Update demo files? (Moved to legacy folder with readme)
- [x] Check if text descriptions near updates from Dist to RV should be added (e.g. if we write stuff like. Define distribution, when we actually now define RV).


24/11/2023
- [x] Whiteboard meeting on RV API

nabriis commented:
Hi @amal-ghamdi and @jakobsj. Following the seminar I have now updated the tutorial. Feel free to review

nabriis commented:
Minor issue if you do 

```python
A = cuqi.model.LinearModel(lambda x: x, lambda y: y, 1, 1))
x = cuqi.distribution.Gaussian(0,1)
y = x+1
A@y # returns x+1
```

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,212 @@
+# Evaluating random variables
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# 
+# Currently, two �modes� of evaluation are supported for random variables:
```
Strange characters around "modes"

jakobsj reviewed (approved):
Quick approval.  Few tiny comments inline.  Looks really good, exciting new functionality. Merge pending more detailed review and approval from @amal-ghamdi

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+        """ Create random variable from distribution """
+        # Convert to list if single distribution
+        if not isinstance(distributions, OrderedSet):
+            distributions = OrderedSet([distributions])
```
I was wondering what is the difference between the OrderedSet here and for example what we use before: "processed_kwargs = set() " in distributions.

nabriis replied:
Ordered set allows the order to be maintained which is e.g. used in Distribution if one of the parameters is a RV and we use positional arguments to condition.

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+                )
+        
+        self._distributions = distributions
+        """ The distribution from which the random variable originates. """
```
Just a question about using docstring here and below, is this preferred over using regular comment?

nabriis replied:
This adds a docstring to the property similar to adding it when using @property.

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+
+
+class RandomVariable:
+    """ Random variable defined by a base distribution with potential algebraic operations applied to it.
```
can you add parameter section for the __init__ parameters `distributions` and `tree`

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+        """ The tree representation of the random variable. """
+
+    def __call__(self, *args, **kwargs) -> Any:
+        """ Evaluate random variable at a given parameter value. """
```
```suggestion
        """ Evaluate random variable at a given parameter value. For example, for random variable `X`, `X(1)` gives `1` and `(X+1)(1)` gives `2` """
```

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+                f"Distributions: \n{parameter_strings}\n"
+                f"Transformations: {self._tree}")
+   
+    def _apply_operation(self, operation, other=None) -> 'RandomVariable':
```
`operation` seems to be a reserved keyword in python. I got a bit confused with the highlight then realized it is an input parameter

nabriis replied:
I do not see this in my IDE. can you give some more details?

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+            return RandomVariable(self._distributions, operation(self._tree))
+        elif isinstance(other, RandomVariable):
+            return RandomVariable(self._distributions | other._distributions, operation(self._tree, other._tree))
+        return RandomVariable(self._distributions, operation(self._tree, other))
```
can you add a comment line mentioning what are the remaining cases that falls under this line

nabriis replied:
Yes. Good suggestion.

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,191 @@
+    def __sub__(self, other) -> 'RandomVariable':
+        return self._apply_operation(operator.sub, other)
+
+    def __rsub__(self, other) -> 'RandomVariable':
```
I think this should use lambda, similar to rtruediv. For example, now I believe `(1-X)(2)` gives `1`

nabriis replied:
Great catch. Thanks :)

amal-ghamdi reviewed (commented):
Thank you @nabriis for this rich feature. Really impressive work you have done here that I enjoyed reviewing and learning from!. 

I have some comments here and there for your consideration.

nabriis commented on cuqi/utilities/_utilities.py:
```diff
@@ -1,4 +1,5 @@
 from cuqi.array import CUQIarray
+import cuqi
```
Yes to avoid circular import we have to use import cuqi here.

nabriis commented on cuqi/distribution/_distribution.py:
```diff
@@ -383,3 +384,59 @@ def __repr__(self) -> str:
+    # The following methods define algebraic operations on distributions
+    # The return type is a RandomVariable instance
+        
+    def as_random_variable(self):
```
Good suggestion

amal-ghamdi replied:
👍

nabriis commented:
Need to decide if `.rv` should even be exposed at this point in time.

nabriis commented on cuqi/distribution/_gmrf.py:
```diff
@@ -99,7 +99,7 @@ def __init__(self, mean, prec, physical_dim=1, bc_type='zero', order=1, is_symme
         super().__init__(is_symmetric=is_symmetric, **kwargs) #TODO: This calls Distribution __init__, should be replaced by calling Gaussian.__init__ 
 
         self.mean = mean.reshape(len(mean), 1)
```
There was some issues that made such a change not backwards compatible. I wanted to keep it out of this PR if we broke backwards compatibility. I created an issue: #258

nabriis commented on cuqi/distribution/_cmrf.py:
```diff
@@ -50,8 +51,8 @@ def __init__(self, location, scale, bc_type="zero", physical_dim=1, **kwargs):
         
-        self.location = location
-        self.scale = scale
+        self.location = if_dist_force_rv(location)
```
I have simplified things here. Merged the force_ndarray pattern and the if_dist_force_rv. Let me know if it is more clear now.

nabriis commented on cuqi/distribution/_gmrf.py:
```diff
@@ -99,7 +99,7 @@ def __init__(self, mean, prec, physical_dim=1, bc_type='zero', order=1, is_symme
 
         self.mean = mean.reshape(len(mean), 1)
-        self.prec = prec
+        self.prec = if_dist_force_rv(prec)
```
I have simplified things here. Merged the force_ndarray pattern and the if_dist_force_rv. Let me know if it is more clear now.

nabriis commented on cuqi/distribution/_cmrf.py:
```diff
@@ -50,8 +51,8 @@ def __init__(self, location, scale, bc_type="zero", physical_dim=1, **kwargs):
         
-        self.location = location
-        self.scale = scale
+        self.location = if_dist_force_rv(location)
```
Yes! I have simplified things here. Merged the force_ndarray pattern and the if_dist_force_rv. Let me know if it is more clear now.

amal-ghamdi replied:
Great!

nabriis commented on cuqi/distribution/_distribution.py:
```diff
@@ -383,3 +384,59 @@ def __repr__(self) -> str:
+        """Return a RandomVariable instance of this distribution"""
+        return self.as_random_variable()
+    
+    def __add__(self, other) -> RandomVariable:
```
Re the first point I liked to keep the code as is to avoid any potential issues with the order of operations etc. Here it is clear that the underlying method is called with the same arguments.

Regarding the 2nd point I created an issue  #259

amal-ghamdi replied:
Makes sense 👍

nabriis commented on cuqi/distribution/_gamma.py:
```diff
@@ -9,8 +10,8 @@ def __init__(self, shape=None, rate=None, is_symmetric=False, **kwargs):
         # Init specific to this distribution
-        self.shape = shape
-        self.rate = rate     
+        self.shape = if_dist_force_rv(shape)
```
Thanks! Fixed.

amal-ghamdi replied:
👍

nabriis commented on cuqi/distribution/_gmrf.py:
```diff
@@ -99,7 +99,7 @@ def __init__(self, mean, prec, physical_dim=1, bc_type='zero', order=1, is_symme
 
         self.mean = mean.reshape(len(mean), 1)
-        self.prec = prec
+        self.prec = if_dist_force_rv(prec)
```
Fixed

nabriis commented on cuqi/distribution/_laplace.py:
```diff
@@ -31,8 +32,8 @@ def __init__(self, location, prec, **kwargs):
 
-        self.location = location
-        self.prec = prec
+        self.location = if_dist_force_rv(location)
```
Fixed

nabriis commented on cuqi/utilities/_utilities.py:
```diff
@@ -9,7 +10,15 @@
+        return dist.as_random_variable()
+    return dist
+
 def force_ndarray(value,flatten=False):
```
I updated the method name and merged it with force_ndarray. It will very likely be useful in other areas than distribution

nabriis commented on cuqi/distribution/_cmrf.py:
```diff
@@ -50,8 +51,8 @@ def __init__(self, location, scale, bc_type="zero", physical_dim=1, **kwargs):
         
-        self.location = location
-        self.scale = scale
+        self.location = if_dist_force_rv(location)
```
Fixed

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,212 @@
+# variables in the following important ways:
+# 
+# 1. They are not immediately evaluated when defined. Instead, they are
+#    evaluated in a �lazy� fashion in the context needed.
```
fixed

nabriis commented on cuqi/randomvariable/_ast.py:
```diff
@@ -0,0 +1,146 @@
+    def __abs__(self):
+        return AbsNode(self)
+
+    def __getitem__(self, i):
```
#259

nabriis commented on cuqi/randomvariable/_ast.py:
```diff
@@ -0,0 +1,146 @@
+    def __rmatmul__(self, other):
+        return MatMulNode(convert_to_node(other), self)    
+
+class RandomVariableNode(Node):
```
Updated documentation.

nabriis commented on cuqi/randomvariable/_ast.py:
```diff
@@ -0,0 +1,146 @@
+        return self.left(**kwargs) / self.right(**kwargs)
+
+class PowerNode(BinaryNodeWithParenthesis):
+    op = '^'
```
Yes. I updated docuimentation further. Hope its more clear now. Check the top docstring.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,212 @@
+# 
+
+BP = BayesianProblem(d, s, x, y)
+print(BP)
```
I added more details to close out the tutorial with the classical example.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,212 @@
+
+
+######################################################################
+# Introduction
```
Good points. I updated tutorial to reflect that. Let me know what you think.

jakobsj replied:
These are excellent point.  We need to think carefully (and I can tell you already have) about the distinction and similarities between random variables and distributions and have the cleanest possible cut. Currently, as Amal points out, in some cases a distribution (`X` in the example) will SOMETIMES behave as a random variable, ie `X+1` makes sense. Other times (ie if used by itself X or X(1)) will behave as a distribution. Different behaviours could be considered, and I am sure you already have so would just like to discuss this, eg., we could require the user to explicitly do `.rv` to go from distribution to random variable before allowing eg addition X+1. Or we could consider ALWAYS operating with random variables without the need for .rv.  For example `X = Gaussian(0,1)` would produce `X` already a random variable, not a distribution. Perhaps could then have `X.dist` or similar than would contain the distribution inside the random variable.

nabriis commented:
Hi @amal-ghamdi and @jakobsj. I have now incorporated all the great comments and suggestions to the PR.

jakobsj reviewed (changes requested):
Brilliant work @nabriis, such a rich new feature, looking forward to getting it in!  I added my preliminary comments to discuss when we meet shortly.

amal-ghamdi commented on cuqi/randomvariable/_ast.py:
```diff
@@ -0,0 +1,227 @@
+class Node:
+    """ Base class for all nodes in the abstract syntax tree.  """
+    def __call__(self, **kwargs):
+        """ Evaluate node at a given parameter value. This will traverse the tree and evaluate given the recorded operations. """
```
```suggestion
        """ Evaluate node at a given parameter value. This will traverse the tree and evaluate it given the recorded operations. """
```

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,204 @@
+    ----------
+    distributions : Distribution or set of Distribution
+        The distribution from which the random variable originates. If multiple distributions are
+        provided, the random variable is defined by the abstract syntax tree representing the
```
```suggestion
        provided, the random variable is defined by the passed abstract syntax `tree` representing the
```

amal-ghamdi commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,204 @@
+    distributions : Distribution or set of Distribution
+        The distribution from which the random variable originates. If multiple distributions are
+        provided, the random variable is defined by the abstract syntax tree representing the
+        algebraic operations on the distributions that define the random variable.
```
suggestion:
```suggestion
        algebraic operations on the random variables defined by the distributions.
```

amal-ghamdi commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,275 @@
+#
+# In the context of CUQIpy, a random variable is a Python object which
+# represents a random variable for the purpose of defining a Bayesian
+# problem. Random variables are defined by their underlying distribution
+# and the algebraic operations performed on them.
```
suggestion, 
```suggestion
# problem. Random variables are defined by their underlying distribution,
# and algebraic operations can be performed on them.
```

amal-ghamdi commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,275 @@
+# 
+# 1. They are not immediately evaluated when defined. Instead, they are
+#    evaluated in a lazy fashion in the context needed.
+# 2. They contain a probability distribution (single or multivariate)
```
```suggestion
# 2. They contain a probability distribution object (single or multivariate)
```

amal-ghamdi reviewed (approved):
Thank you @nabriis for the thorough updates! This looks really great with a vary helpful and illustrative tutorial. I have only few minor comments (mostly minor documentation suggestions for your consideration).

I am approving the PR.

jakobsj commented:
I am working through the tutorial now and enjoying it a lot. Related to my previous comment on `x` vs `X` here are some observations on how a user may be a bit confused about the distinction between the two types.
```
X = Gaussian(0, 1)
```
is a distribution while
```
x = X.rv
```
is a random variable. `x` prints as 
```
RandomVariable
Distributions: 
	X ~ CUQI Gaussian.
Transformations: X
```
the tilde notation I believe is normally used to say a random variable is following a certain distribution, but printed `X` is distribution not random variable, which is `x`. So the printed statement really says that the distribution X follows the CUQI Gaussian distribution, which is not quite right.

We then define 
```
y = (x + 10)**2
```
which prints as 
```
RandomVariable
Distributions: 
	X ~ CUQI Gaussian.
Transformations: (X + 10) ^ 2
```
ie printing about the distribution `X` being added to 10, not the random variable `x`. In fact, `X` was never directly given to `y`, so it is odd that `y` prints it.  This would be clearer if we had chosen to call our random variable `hello = X.rv`, then define `y = (hello + 10)**2`, it still prints with `X`, not `hello`.

Moving on to direct evaluation, I think this gets a bit more counter-intuitive. So we can do
```
y(3)
```
resulting in 
```
169
```
while if we want to provide the keyword/ named variable as in the conditional notation, I would do
```
y(x=3)
```
and expect the same result, since we used `x` to define `y`, but we get
```
100
```
The value 3 appears not to be used, the `X` ignored in the evaluation and simply 10 squared is returned.  On the other hand we can do
```
y(X=3)
```
and get
```
169
```
so this gets the correct answer, but involves asking for the distribution X to be set to 3, which doesn't seem right.

Not completely sure at this point and there may be an easy fix, but currently leaning more toward not allowing the mixing of random variables and distributions, ie, not allowing `X+1` for `X` being a distribution.

jakobsj commented:
Continuing to the part with
```
x = Gamma(1, 1) wh
y = Gaussian(0, 1)
z = x + (y**2+x**2)*0.1
```
this prints as 
```
RandomVariable
Distributions: 
	x ~ CUQI Gamma.
	y ~ CUQI Gaussian.
Transformations: x + (y ^ 2 + x ^ 2) * 0.1
```
which at first looks better, because `x` and `y` are interpreted as random variables when defining `z`, yet they are themselves created as distributions, so the tilde notation still kind of suggests that the "distribution `x` follows a CUQI distribution".

We then go on to evaluate in different variants
```
z(1,2)
```
and 
```
z(x=1, y=2)
```
and even swapping as 
```
z(x=1, y=2)
```
yield
```
1.5
```
so that appears to be good. It is unfortunate that when passing just one of them or random variable names that do not occur in `z` then a result is still produced (I think the value 0 is used for unspecified variables)
```
z(x=1)
```
yields
```
1.1
```
where I'd maybe expect a new random variable with `x` fixed at 1 fixed and still depending on `y`. And 
```
z(hello=9)
```
as well as
```
z(X=1, Y=2)
```
produce
```
0.0
```
where I'd perhaps expect an error or warning - or the unmodified `z` random variable, since no values were passed for the existing `x` and `y` variables that `z` depends on.

jakobsj commented:
I think possibly all this may be clarified if - as we have discussed previously - we change it so that users automatically get a random variable instead of a distribution object.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,275 @@
+
+d = Gamma(1, 1e-4)
+s = Gamma(1, 1e-4)
+x = LMRF(0, 1/d, geometry=A.domain_geometry)
```
I attempted to make it more clear. But I would like to leave the lambda expressions out of this tutorial.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,275 @@
+# transformations performed on it and can use this information to
+# transform the sample drawn from the original distribution.
+# 
+# This means for example that we can both draw a sample from :math:`x` and
```
This is now considered a transformed random variable. With an attribute that indicates if RV is transformed or not.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,275 @@
+#
+# We can achieve this in CUQIpy without invoking `.rv` as follows.
+
+z = Gaussian(0, 1)
```
Updated

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,275 @@
+# We can achieve this in CUQIpy without invoking `.rv` as follows.
+
+z = Gaussian(0, 1)
+v = (z + 10)**2
```
This was updated

nabriis commented on cuqi/distribution/_lmrf.py:
```diff
@@ -75,7 +74,15 @@ def location(self):
     @location.setter
     def location(self, value):
-        self._location = force_ndarray(value, flatten=True)
+        self._location = to_cuqi_format(value, flatten=True)
```
The main motivation to switch from force_ndarray to to_cuqi_format is to generalize the input handling. The arguments will be ignored if you pass e.g. force_ndarray = False and flatten=true. Perhaps we can come up with a better to_cuqi_format method that has more clear parameters.

nabriis commented on cuqi/randomvariable/_ast.py:
```diff
@@ -0,0 +1,227 @@
+
+    x = RandomVariableNode('x')
+    y = RandomVariableNode('y')
+    z = 2*x + 3*y
```
This is a rather technical part. For normal users they wont use this. Its not exposed. In general for two RVs, adding them will create a new RV

nabriis replied:
To answer the question for what the type of z is in this case its of type AddNode. Specifically its 
```
    AddNode(
        MultiplyNode(
            ValueNode(2),
            RandomVariableNode('x')
        ),
        MultiplyNode(
            ValueNode(3),
            RandomVariableNode('y')
        )
    )
```
as written in docstring.

nabriis commented on cuqi/distribution/_lognormal.py:
```diff
@@ -37,8 +38,8 @@ class Lognormal(Distribution):
         super().__init__(is_symmetric=is_symmetric, **kwargs) 
-        self.mean = mean
-        self.cov = cov
+        self.mean = to_cuqi_format(mean, force_ndarray=False)
```
Lets creat an issue to capture this. This is a huge PR already!

nabriis replied:
I believe the Lognormal had this Gaussian to deal with. That is why I left it since it was not "simple"

amal-ghamdi replied:
right! this is even a bigger issue with updating the underlying Gaussian. I created #317

nabriis commented:
> I am working through the tutorial now and enjoying it a lot. Related to my previous comment on `x` vs `X` here are some observations on how a user may be a bit confused about the distinction between the two types.
> 
> ```
> X = Gaussian(0, 1)
> ```
> 
> is a distribution while
> 
> ```
> x = X.rv
> ```
> 
> is a random variable. `x` prints as
> 
> ```
> RandomVariable
> Distributions: 
> 	X ~ CUQI Gaussian.
> Transformations: X
> ```
> 
> the tilde notation I believe is normally used to say a random variable is following a certain distribution, but printed `X` is distribution not random variable, which is `x`. So the printed statement really says that the distribution X follows the CUQI Gaussian distribution, which is not quite right.
> 
> We then define
> 
> ```
> y = (x + 10)**2
> ```
> 
> which prints as
> 
> ```
> RandomVariable
> Distributions: 
> 	X ~ CUQI Gaussian.
> Transformations: (X + 10) ^ 2
> ```
> 
> ie printing about the distribution `X` being added to 10, not the random variable `x`. In fact, `X` was never directly given to `y`, so it is odd that `y` prints it. This would be clearer if we had chosen to call our random variable `hello = X.rv`, then define `y = (hello + 10)**2`, it still prints with `X`, not `hello`.
> 
> Moving on to direct evaluation, I think this gets a bit more counter-intuitive. So we can do
> 
> ```
> y(3)
> ```
> 
> resulting in
> 
> ```
> 169
> ```
> 
> while if we want to provide the keyword/ named variable as in the conditional notation, I would do
> 
> ```
> y(x=3)
> ```
> 
> and expect the same result, since we used `x` to define `y`, but we get
> 
> ```
> 100
> ```
> 
> The value 3 appears not to be used, the `X` ignored in the evaluation and simply 10 squared is returned. On the other hand we can do
> 
> ```
> y(X=3)
> ```
> 
> and get
> 
> ```
> 169
> ```
> 
> so this gets the correct answer, but involves asking for the distribution X to be set to 3, which doesn't seem right.
> 
> Not completely sure at this point and there may be an easy fix, but currently leaning more toward not allowing the mixing of random variables and distributions, ie, not allowing `X+1` for `X` being a distribution.

All of this is now fixed

nabriis commented:
Hey @jakobsj, @amal-ghamdi  and @chaozg. This PR is now ready for the halfway review during the sprint we talked about.

jakobsj reviewed (changes requested):
Hi @nabriis, many thanks for another big iteration on this.  The rv feature is very nice to work with.  I have focused on running through the RandomVariables tutorial and tried out a variety of different use cases I could think of to make sure everything behaves as we would expect, leaving further technical look at the implementation details for later.   Among the things I tried, most appeared consistent.  I did come across a few things that I was surprised about, some of them may be intended behavior while one or two things I think may need looking at, please see specific comments.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+# For example we can evaluate :math:`y` at value ``x=3`` as follows:
+# 
+
+y(3)
```
Fixed! Thanks

nabriis replied:
Yes. Thank you for being thorough. I updated again..

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+y(3)
+
+
+# %%
```
Yes for distributions partial conditioning was possible and still is. This could also be implemented for RVs. I will see if its simple to do.

nabriis replied:
I would suggest we leave this for further exploation in another PR and track this improvement as an issue.

nabriis replied:
#318

nabriis commented on tests/test_joint_distribution.py:
```diff
@@ -283,16 +283,16 @@ def test_MultipleLikelihoodPosterior_should_raise_if_two_densities():
     # Check that we cannot create MultipleLikelihoodPosterior
     with pytest.raises(ValueError, match=r"requires at least three densities"):
-        cuqi.distribution.MultipleLikelihoodPosterior(y.to_likelihood(1), x)
+        cuqi.distribution.MultipleLikelihoodPosterior(y.condition(y=1), x)
```
Good point. I will add more details to the tutorial about these options.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+#    x \sim \mathrm{Gaussian}(0,1)
+# 
+# This can be defined in CUQIpy as follows:
+# 
```
Good suggestion. Added

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+
+y = (x + 10)**2
+
+
```
Added

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+# Evaluating random variables
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# 
+# Currently, two modes of evaluation are supported for random variables:
```
Updated this part to show-case the two main modes

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+
+print(x.sample()) # Draws a sample from x ~ Gaussian(0,1)
+
+print(y.sample()) # Draws a sample from y = (x+10)^2 ~ Gaussian(100,1)
```
I tried to update the explanation. Hope its more clear.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,278 @@
+#
+
+x.logd(3)
+
```
Yes it should have been. This was a bug! Fixed

nabriis commented:
Hello @jakobsj, @amal-ghamdi, @chaozg. I updated the code according to Jakobs comments. I also reverted the to_cuqi_format change, which greatly simplified the additional code to be reviewed (e.g. changes to specific distributions etc.).

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,384 @@
+z = Gaussian(0, lambda s: s).rv
+
+print(z)
+print(z.condition(s=10))
```
Is there a reason to not use the simpler syntax `z(s=10)`, which would be closer to the equivalent for distribution?

nabriis replied:
Yes, the reason is that distributions function function with general "callable" objects in the sense that one can create conditional distributions from using any callable object like a method, a lambda function or an object with the __call__ method implemented.

While we could special case RVs in the distribution implementation, I would rather use this syntax. In fact, we had previously discuseed if we should use a more explicit syntax for conditioning Distributions aswell, rather than the call syntax.

To align RVs and distributions we could thus consider using `.condition` or `.given` as syntax for conditioning. This could be updated in a new PR to avoid keeping this one alive for too long though. See #250

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,384 @@
+# random variable on the given input. This is done using the ``condition``
+# method. For more details see documentation of conditioning on distributions.
+
+z = Gaussian(0, lambda s: s).rv
```
s is used above in the Bayesian problem. I don't think it affects the behaviour, but for clarity use a different symbol not used before, like p

nabriis replied:
Updated.

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,384 @@
+# using the ``logd`` method. **Note**. Only simple (non-transformed)
+# random variables support this method at this point in time.
+#
+
```
```suggestion
# Define point for evaluation
x0 = [1, 2, 3]
```

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,384 @@
+# random variables support this method at this point in time.
+#
+
+x.logd([1, 2, 3])
```
```suggestion
x.logd(x0)
```

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,384 @@
+# probability density function of the random variable at a given point.
+#
+
+x.gradient([1, 2, 3])
```
```suggestion
x.gradient(x0)
```

jakobsj commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,384 @@
+
+print(z)
+print(z.condition(s=10))
+
```
I tried creating the conditioned distribution first, then condition it, then generate the rv, and get an error:
```
Z = Gaussian(0, lambda s: s)
zz = Z(s=3)
zz.rv
```
this is the error
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/.local/lib/python3.9/site-packages/IPython/core/formatters.py in __call__(self, obj)
    700                 type_pprinters=self.type_printers,
    701                 deferred_pprinters=self.deferred_printers)
--> 702             printer.pretty(obj)
    703             printer.flush()
    704             return stream.getvalue()

~/.local/lib/python3.9/site-packages/IPython/lib/pretty.py in pretty(self, obj)
    392                         if cls is not object \
    393                                 and callable(cls.__dict__.get('__repr__')):
--> 394                             return _repr_pprint(obj, self, cycle)
    395 
    396             return _default_pprint(obj, self, cycle)

~/.local/lib/python3.9/site-packages/IPython/lib/pretty.py in _repr_pprint(obj, p, cycle)
    698     """A pprint that just redirects to the normal repr function."""
    699     # Find newlines and replace them with p.break_()
--> 700     output = repr(obj)
    701     lines = output.splitlines()
    702     with p.group():

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in __repr__(self)
    217     def __repr__(self):
    218         # Create strings for parameter name ~ distribution pairs
--> 219         parameter_strings = [f"{name} ~ {distribution}" for name, distribution in zip(self.parameter_names, self.distributions)]
    220         # Join strings with newlines
    221         parameter_strings = "\n".join(parameter_strings)

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in parameter_names(self)
    195     def parameter_names(self) -> str:
    196         """ Name of the parameter that the random variable can be evaluated at. """
--> 197         return [distribution.par_name for distribution in self.distributions]
    198 
    199     @property

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in distributions(self)
    179         # Inject name into distributions
    180         if len(self._distributions) == 1:
--> 181             self._set_dist_name_if_not_set(next(iter(self._distributions)))
    182         return self._distributions
    183 

~/Work/CUQI/dev/CUQIpy/cuqi/randomvariable/_randomvariable.py in _set_dist_name_if_not_set(self, distribution)
    184     def _set_dist_name_if_not_set(self, distribution):
    185         if not hasattr(distribution, '_par_name') or distribution._par_name is None: #TODO. Change to "par_name" once dist does not auto infer
--> 186             distribution.par_name = self.name
    187 
    188     def get_conditioning_variables(self):

~/Work/CUQI/dev/CUQIpy/cuqi/density/_density.py in par_name(self, name)
     45     def par_name(self, name):
     46         if self._is_copy:
---> 47             raise ValueError("Cannot set name of conditioned density. Only the original density can have its name set.")
     48         self._par_name = name
     49 

ValueError: Cannot set name of conditioned density. Only the original density can have its name set.
```

nabriis replied:
I updated such that this can be done now if the original density have not yet a parameter name set. One has to do

```python
Z = Gaussian(0, lambda s: s)
zz = Z(s=3)
z = zz.rv # Name and parameter name becomes z here.
```

jakobsj reviewed (changes requested):
Thank you @nabriis, I've run through the tutorial again and really like the updates you have made. Most points are now addressed. Running through did result in a number of new questions/points, not sure if bugs, please see specific comments. I also resolved many of the conversations. I will take a look at the remaining ones.  I haven't got more time today so I think it makes sense if you can address this round of comments, and any from others, then after that I can take a look again, including at other parts of the code.

jakobsj commented:
I have gone through another time and resolved conversations that had been addressed.  There are four old ones, I think all having to do with if_dist_force_rv.  I think this functionality has changed back and forth to to_cuqi_array or similar, and I'd like to understand these changes and behavior better before resolving these.

nabriis commented:
> I have gone through another time and resolved conversations that had been addressed. There are four old ones, I think all having to do with if_dist_force_rv. I think this functionality has changed back and forth to to_cuqi_array or similar, and I'd like to understand these changes and behavior better before resolving these.

Thanks @jakobsj. I updated once again. In relation to the `force_ndarray` and `to_cuqi_format` these methods are used to have inputs to distributions conform to the expected format. After removing `to_cuqi_format` again, we are currently only forcing ndarray in scalar-type inputs. For example if the user creates a Gaussian like this:

```python
X = Gaussian(0, 1)
```
both the mean and covariance will become ndarrays.

The reason to generalize to `to_cuqi_format` would be to add e.g. error handling for common pitfalls. For example if the user does this:

```python
# A Gaussian distribution with conditioning variable x (from cov) is created in line two:
x = Gaussian(0,1).rv
y = Gaussian(0, x).rv
```

```python
# A Gaussian distribution with a ??Gaussian distribution?? as cov in line two:
x = Gaussian(0,1) # Forgot to put .rv
y = Gaussian(0, x).rv
```

This behavior was also present in current CUQIpy. We could catch this mistake by parsing inputs and if they were a Distribution object, throw error with helpful message asking if the user forgot `.rv`.

amal-ghamdi commented on demos/tutorials/RandomVariables.py:
I experimented a bit with the rvs, and one scenario in inferring the distribution name came to mind (below). Depending on the flag, z2 name will be different. Is that something we want to capture in an issue? Also, is this is something we should prevent by not allowing creating more than one `rv` from a distribution (kind of 1 to 1 association in terms of class relationship).

```
Z_s = cuqi.distribution.Gaussian(0, lambda s: s)
Z = Z_s(s=3)
z = Z.rv
flag = False
if flag:
    print(z.name)
z2 = Z.rv
print(z2.name)
```

nabriis replied:
Good point here. I actually updated now so that .rv creates a copy of the distribution and does NOT introduce side effects to the original variables. Thus `Z_s` and `Z` in the above example does not get their parameter name updated. Instead its updated in `z.dist.par_name`.

amal-ghamdi commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,386 @@
+
+# %%
+######################################################################
+# Direct evaluation
```
May be we can clarify to the user here or in some other place in the tutorial that using () operator means different things when we talk about X and x, X(1) is logpdf and X.rv(1) is the evaluation as described here.

amal-ghamdi reviewed (commented):
Thank you @nabriis again for this carefully and creatively designed PR, powerful and impressive work!

I left some comments throughout, feel free to address or immediately resolve those as it makes sense to you. 

I have few general thoughts for your consideration:
- I thought we can make `par_name` more specific and clear if we call it `rv_name`, if that is applicable in all cases.
- I would assume we need issues to update other demo/plugin repos to address this change.

Well done!

jakobsj commented:
I believe I have now understood the points about forcing rv, etc, and I agree that in the spirit of being clear about when something is an RV and a distb respectively, we should not convert automatically between the two behind the scenes. Instead, as you suggest suitable error messages should be given when a user tries to mix.

jakobsj commented:
I have spent the morning thinking more about distributions, random variables and the "call" syntax.  I am not sure if this is the best place to capture these, but for now I will put them here.  Please note, that I honestly feel the random variable is an amazing addition, and I am only giving it so much scrutiny, because it introduces such a fundamental change in the usage of CUQIpy, that we want to make sure everything remains (or becomes!) consistent.   

This will be the first part, outlining some thoughts first on "call" syntax with only distributions, followed by using only random variables.  I expect to follow up later (but out of time for now), with additional thoughts on potential cases of combining the two. Sorry if it reads slightly rambling at times, but I thought I'd sent it now, rather than let it sit until I have time to streamline it more.

Only distributions

Before introduction of RVs, we have distributions 

X = Gaussian(0,1)

One can do 

X(4)
EvaluatedDensity([-8.91893853])

in which the “call” syntax amounts to returning the scalar value of the pdf at the input value.

We can have a “simple” conditional distribution but eg omitting an input

Y = Gaussian(0)
Y(2)     # or Y(cov=2)
CUQI Gaussian.

In which the “call” syntax amounts to conditioning on cov=2, and returning the resulting (unconditional) distribution.

Similarly using lambda function can set up more involved conditional distributions such as
Z = Gaussian(lambda t: 1/t)
Z(4)
CUQI Gaussian.

Two issues here:

1. The “call” syntax means two different things, depending on whether used with an unconditional or a conditional distribution.  In the extreme, one can in fact employ both in a single call
Y(4)(3)
which first conditions on cov=4, then evaluates the pdf at 3. If one knows what is going on, this can be understood, but is probably very confusing to users.

2. For the conditional distribution with lambda function, it is unclear what
Z(4)
refers to, cov or t. Turns out one can in fact call with either! Followed by evaluation, we can see we get different results:
Z(cov=4)(0)
EvaluatedDensity([-1.61208571])

Z(t=4)(0)
EvaluatedDensity([-0.22579135])

and doing this without keyword we see that Z(4)  corresponds to Z(t=4):
Z(4)(0)
EvaluatedDensity([-0.22579135])
as we get the same value.

It surprises me that using “cov” is permitted, when we have explicitly made Y conditional on t, instead of cov. Is that what we want?

Ideas to consider here: Replace both call syntax by more explicit syntax. And consider disallowing the usage of “cov” in conditioning when lambda function is used. And/or disallow positional inputs, allowing only keyword based for clarity.


Only random variables:


Introducing random variables (RVs):

X = Gaussian(0, 1)
x = X.rv

we can now use “call” syntax (of the RV) to do
x(4)     # or x(x=4)
producing 
4
because this is “direct evaluation”, meaning we fix x at 4, which means the value of x is indeed 4, ignoring that it also happens to be a Gaussian random variable. 

For a transformed RV, eg.
y = (x + 10)**2
using the “call” syntax is consistent with this (and a bit clearer), since it also simply fixes the value of x, e.g., 
y(x=3)
produces 169, again ignoring that x is a Gaussian random variable.

Conceptually, this appears very similar to conditioning for distributions, in the sense that we fix one and given that evaluate the other. Except the transformed variable here is doing algebraic functions, not feeding through a distribution.  But we do want to use RVs to specify relations between Rvs/distributions, not just algebraic manipulations. So we want to do eg
a = Gaussian(0,1).rv
b = Gaussian(0, 1/a).rv
which makes both RVs:
a ~ CUQI Gaussian.
b ~ CUQI Gaussian. Conditioning variables ['a'].
What happens if we use call syntax?
a(5)    # or  a(a=5)
5    which is consistent with the x above.
Same for b, which completely disregards that it is conditional on a:
b(5)  # or b(b=5):
5
We get an error if we try
b(a=5)
ValueError: Expected arguments ['b'], got arguments {'a': 5}
So we cannot use “call” syntax to “directly evaluate” b for a fixed value of a, as in the case of y given a value of x above.  This is probably because y only used algebraic operations, while b is a conditional distribution w.r.t. a.  Currently, one then needs to do 
b.condition(a=5)
yielding
b ~ CUQI Gaussian.
So now it has been conditioned on a. Maybe this usage is indeed not “direct evaluation”, since we end up with a fully specified distribution, not a value as in direct evaluation.  But in general, it seems a subtle difference, and I could imagine more involved combinations of algebraic expressions and conditional distribution RVs for defining an RV. As well as wanting to call b.sample() and expect it to draw a sample from a, then use that in b. That gives an error at present, but mathematically nothing prevents us from expecting this to work.

An additional aspect is that before conditioning we have 
print(b)
b ~ CUQI Gaussian. Conditioning variables ['a'].
but above, and also if we assign to, say c, we still get “b” printed:
c = b.condition(a=5)
print(c)
b ~ CUQI Gaussian.

nabriis commented on demos/tutorials/RandomVariables.py:
```diff
@@ -0,0 +1,386 @@
+# Providing data to the Bayesian problem creates a posterior as shown.
+#
+
+BP.set_data(y=y_obs)
```
I can't recall. But yes its not a "Posterior" class, but a JointDistribution representing a posterior over joint set of variables

nabriis commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,337 @@
+    def _make_copy(self):
+        """ Returns a shallow copy of the density keeping a pointer to the original. """
+        new_variable = copy(self)
+        new_variable._distributions = copy(self.distributions)
+        new_variable._tree = copy(self._tree)
```
Not, it is a shallow copy. So copying here ensures we get a new pointer to the distributions, meaning if we change internal parameters it only affects this new copied RV.

nabriis replied:
Think of it as a 2-level deep copy.

nabriis commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,337 @@
+    def get_conditioning_variables(self):
+        """ Get conditioning variables. """
+        if self.is_transformed:
+            raise NotImplementedError("Extracting conditioning variables is not implemented for transformed random variables")
```
I believe we should generate a tracking issue to track all of the features we have disabled and aim to create in the future for RV class. There are more than thus, but if we create the issue we can add bullets to it.

nabriis commented on cuqi/distribution/_distribution.py:
```diff
@@ -429,3 +430,16 @@ def __repr__(self) -> str:
         else:
             return "CUQI {}.".format(self.__class__.__name__)
+        
+    # The following methods define algebraic operations on distributions
+    # The return type is a RandomVariable instance
```
Good catch! Removed.

nabriis commented on cuqi/likelihood/_likelihood.py:
```diff
@@ -35,13 +35,22 @@ def __init__(self, distribution, data):
+    
+    @_par_name.setter
+    def _par_name(self, value):
+        self.distribution._par_name = value
```
Good catch. This is now removed and it helped simplify some checks in RV class, which checked for the private property _par_name where it should not have done so.

nabriis commented on cuqi/likelihood/_likelihood.py:
```diff
@@ -163,24 +172,24 @@ class UserDefinedLikelihood(object):
-            self._name = _get_python_variable_name(self)
-        return self._name
+        if self._par_name is None:
+            self._par_name = _get_python_variable_name(self) # TODO
```
Its related. Currently the UserDefinedLikelihood can infer python variable names where normally after this change it was supposed to only be RV class that does that. However, we have some use cases, where users define their UserDefinedLikelihood using e.g.
y = UserDefinedLikelihood(...)

Assuming it gets the name y. I think we can add this info into the issue you mentioned.

nabriis commented on cuqi/model/_model.py:
```diff
@@ -336,8 +337,16 @@ def forward(self, *args, is_par=True, **kwargs):
+        if isinstance(x, cuqi.randomvariable.RandomVariable):
+            return self._handle_random_variable(x)
+        
+        # If input is a Node from internal abstract syntax tree, we let the Node handle the operation
```
Actually this is not the same as raising a NotImplementedError. 

This is a signalling tool to indicate that `Model` does not have the operation implemented, but without raising an error just yet. This is usueful e.g. for stuff like:

```python
x = Gaussian(0,1).rv
y = A@(x+1)
```

Here by convention, Python will first request `A` for its `__matmul__` implementation. If it is implemented it will call it. If not it calls `__rmatmul` of (x+1).

In `model` we only support non-transformed random variables as we can change the name of the input parameter in that case. For more complicated cases, we have to rely on the general implementation of the RV class.

So, for Transformed Random Variables like the (x+1) above, we want to return `NotImplemented` to indicate that `x+1` should handle the matrix multiplication usings its `__rmatmul__` method.

nabriis commented on cuqi/utilities/_get_python_variable_name.py:
```diff
@@ -9,7 +9,7 @@
     """ Retrieve the Python variable name of an object. Takes the first variable name appearing on the stack that is not in the ignore list. """
 
-    ignored_var_names = ["self", "cls", "obj", "var", "_"]
+    ignored_var_names = ["self", "cls", "obj", "var", "_", "result", "args", "kwargs", "par_name", "name", "distribution", "dist"]
```
A warning will be raised saying the name `dist` is reserved

nabriis commented on cuqi/randomvariable/_ast.py:
```diff
@@ -0,0 +1,227 @@
+    def __repr__(self):
+        return f"abs({self.child})"
+    
+class MatMulNode(BinaryNodeWithParenthesis):
```
It uses the repr of BinaryNodeWithParenthesis

nabriis commented on cuqi/randomvariable/_orderedset.py:
```diff
@@ -0,0 +1,59 @@
+    """A set (i.e. unique elements) that keeps its elements in the order they were added.
+
+    This is a minimal implementation of an ordered set, using a dictionary for storage.
+    """
```
Yes - as you wrote its always None with item stored as key. I am sure there could be other ways to do this, but since its an internal structure within the RV this approach is OK for me.

nabriis commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,337 @@
+        d = Gamma(1, 1e-4)
+        s = Gamma(1, 1e-4)
+        x = GMRF(np.zeros(A.domain_dim), d)
+        y = Gaussian(A @ x, 1/s)
```
Thanks. I had forgotten to update this part. You must create the random variable now using `.rv`. Updated examples.

nabriis commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,337 @@
+    @name.setter
+    def name(self, name):
+        if self._is_copy:
+            raise ValueError("Cannot set name of conditioned random variable. Only the original variable can have its name set.")
```
Fixed.

nabriis commented on cuqi/randomvariable/_randomvariable.py:
```diff
@@ -0,0 +1,337 @@
+        return distribution
+    
+    @property
+    def distributions(self) -> set:
```
Dist is used to extract a single distribution for a simple random variable. Its short and nice for use e.g. with `x.dist.rv.dist` etc. :)

Perhaps `Distributions` should just be a private attribute? Its not really supposed to be user facing.
